### ISSUE TITLE
A callback holded in another callback never execute

### ISSUE THREAD

**Issue by zhouyibing** [OpenTSDB/asynchbase#175]
byte[] table = Bytes.toBytes("device");
        byte[] k = BytesUtil.toBytes("abcde");
        byte[] f = BytesUtil.toBytes("f");
        byte[] f2 = BytesUtil.toBytes("f2");
        byte[] q = BytesUtil.toBytes("dd");
        final GetRequest req = new GetRequest(table,k,f);
        final GetRequest req2 = new GetRequest(table,k,f2);
        final HBaseClient client = new HBaseClient("10.2.4.190,10.2.4.189,10.2.4.191","/hbase");
        Deferred<ArrayList<KeyValue>> ret = client.get(req);
        ret.addCallback(new Callback<Object, ArrayList<KeyValue>>() {
            @Override
            public Object call(ArrayList<KeyValue> arg) throws Exception {
                System.out.println("cccc");
                client.get(req2).addCallback(new Callback<Object, ArrayList<KeyValue>>() {
                    @Override
                    public Object call(ArrayList<KeyValue> arg) throws Exception {
                        System.out.println("ggg");
                        return arg;
                    }
                });
                System.out.println("ddddd");
                return arg;
            }
        });
        Thread.currentThread().join();

**Comment by jasonculverhouse** [comment]
your `Thread.currentThread().join();` should more likely be `ret.join();`

If you want to wait on both to complete then You have 2 choices:

if the second get is dependent on the results of the first get then you connect them with `addCallBackDeferring` and the result is a `Deferred` that completes after both complete "serially"

    Deferred<ArrayList<KeyValue>> ret = client.get(req)
            .addCallbackDeferring(new Callback<Deferred<ArrayList<KeyValue>>, ArrayList<KeyValue>>() {
                @Override
                public Deferred<ArrayList<KeyValue>> call(ArrayList<KeyValue> arg) throws Exception {
                    System.out.println("cccc");
                    return client.get(req2).addCallback(new Callback<ArrayList<KeyValue>, ArrayList<KeyValue>>() {
                        @Override
                        public ArrayList<KeyValue> call(ArrayList<KeyValue> arg) throws Exception {
                            System.out.println("ggg");
                            return arg;
                        }
                    });
                }
            });
    ret.join();

if these are truly parallel requests then what you want to do is execute them both in parallel and wait on the grouped result

    Deferred<ArrayList<ArrayList<KeyValue>>> ret = Deferred.group(
            client.get(req).addCallback(new Callback<ArrayList<KeyValue>, ArrayList<KeyValue>>() {
                @Override
                public ArrayList<KeyValue> call(ArrayList<KeyValue> arg) throws Exception {
                    System.out.println("cccc");
                    return arg;
                }
            }),
            client.get(req2).addCallback(new Callback<ArrayList<KeyValue>, ArrayList<KeyValue>>() {
                @Override
                public ArrayList<KeyValue> call(ArrayList<KeyValue> arg) throws Exception {
                    System.out.println("ggg");
                    return arg;
                }
            }));

    ret.join();

**Comment by IDerr** [comment]
@manolama you can close this one
